Actually there are several "filter suite" regressions with this change:

FAILED filter_with_non_root_user_test.py::test_all_positive[(&(!(l=Cupertino))(!(|(uid=*wal*)(&(sn~=tiller) (roomNumber=2295)))))] - AssertionError: assert []
FAILED filter_with_non_root_user_test.py::test_all_positive[(&(l=Cupertino)(!(|(uid=*wal*)(&(sn~=tiller) (roomNumber=2295)))))] - AssertionError: assert []
FAILED filter_with_non_root_user_test.py::test_all_positive[(!(|(!(l=*))(!(l=sunnyvale))))] - AssertionError: assert []
FAILED filter_with_non_root_user_test.py::test_all_positive[(&(!(l=Cupertino))(!(mail=*exam*))(!(|(uid=*wal*) (l=*))))] - AssertionError: assert []
FAILED large_filter_test.py::test_large_filter[(&(objectClass=person)(|(manager=uid=fmcdonnagh *)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_0,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_1,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_2,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_3,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_4,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_5,*)(manager=uid=jvedder,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_6,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_7,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_8,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_9,*)(manager=cn=no_such_entry_with_a_really_long_dn_component_to_stress_the_filter_handling_code_10,*)(manager=uid=cnewport,*)))]
FAILED vfilter_simple_test.py::test_param_positive[(&(!(l=Cupertino))(!(|(uid=*wal*)(&(nsRole~=cn=new managed) (nsRole=cn=*vaddr*)))))] - AssertionError: assert []
FAILED vfilter_simple_test.py::test_param_positive[(&(l=Cupertino)(!(|(uid=*wal*)(&(nsRole~=cn=new managed) (nsRole=cn=*vaddr*)))))] - AssertionError: assert []
FAILED vfilter_simple_test.py::test_param_positive[(!(|(!(l=*))(!(l=sunnyvale))))] - AssertionError: assert []
FAILED vfilter_simple_test.py::test_param_positive[(&(!(l=Cupertino))(!(emailclass=*emai*))(!(|(nsRole=cn=*vaddr*) (l=*))))] - AssertionError: assert []

--- a/dirsrvtests/tests/suites/filter/filter_test_aci_with_optimiser.py
+++ b/dirsrvtests/tests/suites/filter/filter_test_aci_with_optimiser.py
@@ -1,0 +1,91 @@
+
+
+import ldap
+import logging
+import pytest
+import os
+from lib389._constants import *
+from lib389.topologies import topology_st as topo
+from lib389.idm.domain import Domain
+from lib389.idm.organizationalunit import OrganizationalUnits
+from lib389.idm.account import Anonymous
+
+log = logging.getLogger(__name__)
+
+
+def test_filter_access(topo):
+    """Search that compound filters are correctly processed by access control
+
+    :id: ad6a3ffc-2620-4e76-909b-926f94c1a920
+    :setup: Standalone Instance
+    :steps:
+        1. Add anonymous aci
+        2. Add ou
+        2. Test good filters
+        4. Test bad filters
+    :expectedresults:
+        1. Success
+        2. Success
+        3. The good filters return the OU entry
+        4. The bad filters do not return the OU entry
+    """
+
+    # Add aci
+    ACI_TEXT = ('(targetattr="objectclass || cn")(version 3.0; acl "Anonymous read access"; allow' +
+                '(read, search, compare) userdn = "ldap:///anyone";)')
+    domain = Domain(topo.standalone, DEFAULT_SUFFIX)
+    domain.replace('aci', ACI_TEXT)
+
+    # To remove noise, delete EVERYTHING else.
+
+    ous = OrganizationalUnits(topo.standalone, DEFAULT_SUFFIX)
+    existing_ous = ous.list()
+    for eou in existing_ous:
+        eou.delete(recursive=True)
+
+    # Create restricted entry
+    OU_PROPS = {
+        'ou': 'restricted',
+        'description': 'secret data'
+    }
+    ou = ous.create(properties=OU_PROPS)
+    OU_DN = ou.dn
+
+    # Do anonymous search using different filters
+    GOOD_FILTERS = [
+        "(|(objectClass=top)(&(objectClass=organizationalunit)(description=secret data)))",
+        "(|(&(objectClass=organizationalunit)(description=secret data))(objectClass=top))",
+        "(|(objectClass=organizationalunit)(description=secret data)(sn=*))",
+        "(|(description=secret data)(objectClass=organizationalunit)(sn=*))",
+        "(|(sn=*)(description=secret data)(objectClass=organizationalunit))",
+        "(objectClass=top)",
+    ]
+    BAD_FILTERS = [
+        "(|(objectClass=person)(&(objectClass=organizationalunit)(description=secret data)))",
+        "(&(objectClass=top)(objectClass=organizationalunit)(description=secret data))",
+        "(|(&(description=*)(objectClass=top))(objectClass=person))",
+        "(description=secret data)",
+        "(description=*)",
+        "(ou=*)",
+    ]
+    conn = Anonymous(topo.standalone).bind()
+
+    # These searches should return the OU
+    for search_filter in GOOD_FILTERS:
+        entries = conn.search_s(OU_DN, ldap.SCOPE_SUBTREE, search_filter)
+        log.debug(f"Testing good filter: {search_filter} result: {len(entries)}")
+        assert len(entries) == 1
+
+    # These searches should not return the OU
+    for search_filter in BAD_FILTERS:
+        entries = conn.search_s(OU_DN, ldap.SCOPE_SUBTREE, search_filter)
+        log.debug(f"Testing bad filter: {search_filter} result: {len(entries)}")
+        assert len(entries) == 0
+
+
+if __name__ == '__main__':
+    # Run isolated
+    # -s for DEBUG mode
+    CURRENT_FILE = os.path.realpath(__file__)
+    pytest.main(["-s", CURRENT_FILE])
+

--- a/ldap/servers/slapd/back-ldbm/ldbm_search.c
+++ b/ldap/servers/slapd/back-ldbm/ldbm_search.c
@@ -1723,9 +1723,9 @@ ldbm_back_next_search_entry(Slapi_PBlock *pb)
                          * we have messed with internally - remember, our internal changes are secure!
                          */
                         slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry",
-                                      "Bypassing filter test\n");
+                                      "Bypassing filter test for %s\n", slapi_entry_get_dn_const(e->ep_entry));
                         if (ACL_CHECK_FLAG) {
-                            filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
+                            filter_test = slapi_vattr_filter_test(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG);
                         } else {
                             filter_test = 0;
                         }
@@ -1757,13 +1757,19 @@ ldbm_back_next_search_entry(Slapi_PBlock *pb)
                          * we need to STILL apply the filter test with "as executed" in case of a test threshold
                          * shortcut (lest we accidentally prevent the user seeing what they wanted ....)
                          */
-                        filter_test = slapi_vattr_filter_test_ext(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG, 1 /* Only perform access checking, thank you */);
+                        slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry",
+                                      "Applying filter test to %s\n", slapi_entry_get_dn_const(e->ep_entry));
+                        filter_test = slapi_vattr_filter_test(pb, e->ep_entry, filter_intent, ACL_CHECK_FLAG);
+                        slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry",
+                                      "Applying filter test intermediate value %d \n", filter_test);
                         if (filter_test == 0) {
                             filter_test = slapi_vattr_filter_test(pb, e->ep_entry, filter, 0);
                         }
                     }
                 }
             }
+            slapi_log_err(SLAPI_LOG_FILTER, "ldbm_back_next_search_entry",
+                          "filter test value %d %s \n", filter_test, slapi_entry_get_dn_const(e->ep_entry));
             if ((filter_test == 0) || (sr->sr_virtuallistview && (filter_test != -1)))
             /* ugaston - if filter failed due to subentries or tombstones (filter_test=-1),
              * just forget about it, since we don't want to return anything at all. */

--- a/ldap/servers/slapd/filterentry.c
+++ b/ldap/servers/slapd/filterentry.c
@@ -764,6 +764,12 @@ slapi_vattr_filter_test_ext(
     int rc = 0; /* a no op request succeeds */
     int access_check_done = 0;
 
+    if (only_check_access != 0) {
+        slapi_log_err(SLAPI_LOG_ERR, "slapi_vattr_filter_test_ext",
+            "⚠️  DANGER ⚠️  - only_check_access mode is BROKEN!!! YOU MUST CHECK ACCESS WITH FILTER MATCHING");
+    }
+    PR_ASSERT(only_check_access == 0);
+
     /* Fix for ticket 48275
      * If we want to handle or components which can contain nonmatching components without access propoerly
      * always filter verification and access check have to be done together for each component
@@ -940,6 +946,9 @@ test_filter_access(Slapi_PBlock *pb,
     rc = plugin_call_acl_plugin(pb, e, attrs, attr_val,
                                 SLAPI_ACL_SEARCH, ACLPLUGIN_ACCESS_DEFAULT, NULL);
 
+    slapi_log_err(SLAPI_LOG_FILTER, "slapi_vattr_filter_test_ext_internal",
+                  "acl result for %s %s = %d\n", slapi_entry_get_dn_const(e), attr_type, rc);
+
     return (rc);
 }
 
@@ -1013,14 +1022,17 @@ vattr_test_filter_list_or(
                 continue;
             }
         }
-        if (only_check_access)
-            continue;
         /* now check if filter matches */
+        /*
+         * We can NOT skip this because we need to know if the item we matched on
+         * is the item with access denied.
+         */
         undefined = 0;
         rc = slapi_vattr_filter_test_ext_internal(pb, e, f, 0, 0, access_check_done);
         if (rc == 0) {
             undefined = 0;
             nomatch = 0;
+            /* We matched, and have access. we can now return */
             break;
         } else if (rc > 0) {
             undefined = rc;

